File: adapters/go-stdlib/cmd/adapter/discovery.go
package main

import (
	"bytes"
	"context"
	"errors"
	"sort"
)

const (
	discoveryVersion          = "1.0.0"
	discoveryRequestSchemaURI = "urn:ucf:adapter:discovery-request:1.0.0"
	discoveryResultSchemaURI  = "urn:ucf:adapter:discovery-result:1.0.0"
	discoveryProcedureURI     = "urn:ucf:onboarding-procedure:go-stdlib-static-discovery:1.0.0"
	discoveryConfidenceBasis  = "urn:ucf:onboarding-confidence:go-exported-function:1.0.0"
)

type discoveryPortSpec struct {
	name      string
	valueKind string
	required  bool
}

type discoveryFunctionSpec struct {
	slug       string
	confidence string
	inputs     []discoveryPortSpec
	outputs    []discoveryPortSpec
}

var discoveryFunctions = map[string]discoveryFunctionSpec{
	"FormatReceipt": {
		slug:       "format-receipt",
		confidence: "0.82",
		inputs: []discoveryPortSpec{
			{name: "total-cents", valueKind: "integer", required: true},
		},
		outputs: []discoveryPortSpec{
			{name: "receipt", valueKind: "string", required: true},
		},
	},
	"LegacyDiscountHint": {
		slug:       "legacy-discount-hint",
		confidence: "0.61",
		inputs: []discoveryPortSpec{
			{name: "code", valueKind: "string", required: true},
		},
		outputs: []discoveryPortSpec{
			{name: "discount-percent", valueKind: "integer", required: false},
		},
	},
	"NormalizeCoupon": {
		slug:       "normalize-coupon",
		confidence: "0.82",
		inputs: []discoveryPortSpec{
			{name: "code", valueKind: "string", required: true},
		},
		outputs: []discoveryPortSpec{
			{name: "normalized-code", valueKind: "string", required: true},
		},
	},
	"QuoteOrder": {
		slug:       "quote-order",
		confidence: "0.82",
		inputs: []discoveryPortSpec{
			{name: "quantity", valueKind: "integer", required: true},
			{name: "unit-price-cents", valueKind: "integer", required: true},
		},
		outputs: []discoveryPortSpec{
			{name: "total-cents", valueKind: "integer", required: true},
		},
	},
}

func (instance *server) discovery(
	contextValue context.Context,
	payload map[string]any,
) operationOutcome {
	run := instance.inventoryRun
	if run == nil || !run.completed {
		return failedOperation(
			"operation_failed",
			"discovery requires a completed inventory snapshot",
		)
	}
	request, err := decodeDiscoveryRequest(payload, run)
	if err != nil {
		return failedOperation(
			"invalid_params",
			"discovery request is outside the Go fixture profile",
		)
	}
	if contextValue.Err() != nil {
		return cancelledOperation()
	}
	result, err := instance.buildDiscoveryResult(request, run)
	if err != nil {
		return failedOperation(
			"internal_error",
			"discovery result construction failed",
		)
	}
	tagged, ok := encodeProfileValue(result, 0)
	if !ok {
		return failedOperation(
			"internal_error",
			"discovery result cannot be encoded",
		)
	}
	if contextValue.Err() != nil {
		return cancelledOperation()
	}
	return successfulOperation(map[string]any{
		"kind": "discover_result",
		"payload": map[string]any{
			"kind":           "adapter_payload",
			"schema_uri":     discoveryResultSchemaURI,
			"schema_version": discoveryVersion,
			"value":          tagged,
		},
	})
}

func decodeDiscoveryRequest(
	payload map[string]any,
	run *inventoryRun,
) (map[string]any, error) {
	if payload["schema_uri"] != discoveryRequestSchemaURI ||
		payload["schema_version"] != discoveryVersion {
		return nil, errors.New("incompatible discovery payload")
	}
	logical, err := decodeProfileValue(payload["value"], 0)
	if err != nil {
		return nil, err
	}
	request, ok := logical.(map[string]any)
	if !ok || !hasExactKeys(
		request,
		"capability",
		"inventory",
		"inventory_binding",
		"kind",
		"onboarding_version",
		"schema_uri",
	) ||
		request["kind"] != "discovery_request_profile" ||
		request["onboarding_version"] != discoveryVersion ||
		request["schema_uri"] != discoveryRequestSchemaURI ||
		!sameCanonicalLogicalValue(
			request["capability"],
			discoveryCapability(),
		) {
		return nil, errors.New("invalid discovery request")
	}
	embedded, ok := request["inventory"].(map[string]any)
	if !ok || !sameCanonicalLogicalValue(embedded, run.snapshot) {
		return nil, errors.New("discovery inventory does not match active run")
	}
	if !validDiscoveryInventoryBinding(
		request["inventory_binding"],
		run,
	) {
		return nil, errors.New("discovery inventory binding is stale")
	}
	recomputed, err := canonicalLogicalDigest(run.snapshot)
	if err != nil || recomputed != run.snapshotDigest ||
		!sameCanonicalLogicalValue(run.records, run.snapshot["records"]) {
		return nil, errors.New("active inventory run is invalid")
	}
	return request, nil
}

func validDiscoveryInventoryBinding(value any, run *inventoryRun) bool {
	binding, ok := value.(map[string]any)
	if !ok || !hasExactKeys(
		binding,
		"canonical_digest",
		"inventory_version",
		"kind",
		"schema_uri",
		"source_revision",
		"subject_uri",
	) {
		return false
	}
	expected := map[string]any{
		"kind":              "inventory_binding",
		"schema_uri":        inventorySchemaURI,
		"inventory_version": inventoryVersion,
		"subject_uri":       run.snapshot["subject_uri"],
		"source_revision":   run.snapshot["source_revision"],
		"canonical_digest":  digestValue(run.snapshotDigest),
	}
	return sameCanonicalLogicalValue(binding, expected)
}

func (instance *server) buildDiscoveryResult(
	request map[string]any,
	run *inventoryRun,
) (map[string]any, error) {
	interfaces := make([]map[string]any, 0, 12)
	manifests := make([]map[string]any, 0, 1)
	for _, raw := range run.records {
		record, ok := raw.(map[string]any)
		if !ok {
			return nil, errors.New("inventory record is invalid")
		}
		switch record["kind"] {
		case "build_manifest":
			manifests = append(manifests, record)
		case "public_interface":
			interfaces = append(interfaces, record)
		}
	}
	supportedFunctions := discoveryFunctions
	expectedInterfaceCount := 12
	expectedCandidateCount := len(discoveryFunctions)
	if instance.platformMode {
		supportedFunctions = map[string]discoveryFunctionSpec{
			"QuoteOrder": discoveryFunctions["QuoteOrder"],
		}
		expectedInterfaceCount = 9
		expectedCandidateCount = 1
	}
	if len(interfaces) != expectedInterfaceCount || len(manifests) != 1 {
		return nil, errors.New("inventory discovery facts are incomplete")
	}
	sortInventoryRecords(interfaces)
	sortInventoryRecords(manifests)
	if manifests[0]["dialect_uri"] != goModuleDialectURI {
		return nil, errors.New("inventory manifest is unsupported")
	}

	context := map[string]any{
		"inventory_binding": request["inventory_binding"],
		"producer":          inventoryProducer(),
		"capability":        discoveryCapability(),
		"procedure_uri":     discoveryProcedureURI,
	}
	candidates := make([]any, 0, expectedCandidateCount)
	eligible := make([]any, 0, len(interfaces))
	uncovered := make([]any, 0, len(interfaces)-len(discoveryFunctions))
	for _, subject := range interfaces {
		reference, err := inventoryReferenceForDiscovery(subject)
		if err != nil {
			return nil, err
		}
		eligible = append(eligible, reference)
		name, nameOK := subject["name"].(string)
		specification, supported := supportedFunctions[name]
		if !nameOK || !supported ||
			subject["interface_kind_uri"] != goExportedFunctionURI {
			uncovered = append(uncovered, reference)
			continue
		}
		proposal := discoveryProposal(specification)
		semanticDigest, err := canonicalLogicalDigest(proposal)
		if err != nil {
			return nil, err
		}
		evidence := []any{reference}
		if name == "QuoteOrder" {
			evidence, err = instance.quoteOrderImplementationEvidence(
				run.records,
			)
			if err != nil {
				return nil, err
			}
		}
		candidate := map[string]any{
			"kind":            "discovery_candidate",
			"id":              "candidate." + zeroDigest(),
			"semantic_digest": digestValue(semanticDigest),
			"subject":         reference,
			"evidence":        evidence,
			"confidence": map[string]any{
				"kind":  "confidence",
				"scale": "decimal-0-to-1",
				"value": specification.confidence,
				"basis": discoveryConfidenceBasis,
			},
			"proposal": proposal,
		}
		candidateID, err := discoveryCandidateID(candidate, context)
		if err != nil {
			return nil, err
		}
		candidate["id"] = "candidate." + candidateID
		candidates = append(candidates, candidate)
	}
	sortDiscoveryReferences(eligible)
	sortDiscoveryReferences(uncovered)
	sort.Slice(candidates, func(left int, right int) bool {
		return candidates[left].(map[string]any)["id"].(string) <
			candidates[right].(map[string]any)["id"].(string)
	})
	if len(candidates) != expectedCandidateCount ||
		len(uncovered) != len(interfaces)-expectedCandidateCount {
		return nil, errors.New("discovery coverage is inconsistent")
	}
	return map[string]any{
		"kind":               "discovery_result_profile",
		"onboarding_version": discoveryVersion,
		"schema_uri":         discoveryResultSchemaURI,
		"inventory_binding":  context["inventory_binding"],
		"producer":           context["producer"],
		"capability":         context["capability"],
		"procedure_uri":      context["procedure_uri"],
		"coverage": map[string]any{
			"kind":               "discovery_coverage",
			"status":             "partial",
			"eligible_subjects":  eligible,
			"uncovered_subjects": uncovered,
		},
		"diagnostics": []any{},
		"candidates":  candidates,
	}, nil
}

func discoveryProposal(specification discoveryFunctionSpec) map[string]any {
	slug := specification.slug
	actionID := "action." + slug
	stepID := "step." + slug
	useCaseID := "use-case." + slug
	actionReference := discoveryEntityReference("proposed_action", actionID)
	stepReference := discoveryEntityReference("proposed_step", stepID)
	useCaseReference := discoveryEntityReference("proposed_use_case", useCaseID)
	inputs := discoveryPorts(specification.inputs)
	outputs := discoveryPorts(specification.outputs)
	bindings := make([]any, 0, len(inputs))
	bindingReferences := make([]any, 0, len(inputs))
	for _, raw := range inputs {
		input := raw.(map[string]any)
		name := input["name"].(string)
		bindingID := "binding." + slug + "." + name
		bindingReference := discoveryEntityReference(
			"proposed_binding",
			bindingID,
		)
		bindingReferences = append(bindingReferences, bindingReference)
		bindings = append(bindings, map[string]any{
			"kind": "proposed_binding",
			"id":   bindingID,
			"target": map[string]any{
				"kind":      "proposal_port_ref",
				"owner":     stepReference,
				"direction": "input",
				"name":      name,
			},
			"source": map[string]any{
				"kind":      "proposal_port_ref",
				"owner":     useCaseReference,
				"direction": "input",
				"name":      name,
			},
		})
	}
	entities := []any{
		map[string]any{
			"kind":         "proposed_action",
			"id":           actionID,
			"input_ports":  inputs,
			"output_ports": outputs,
		},
	}
	entities = append(entities, bindings...)
	entities = append(
		entities,
		map[string]any{
			"kind":     "proposed_step",
			"id":       stepID,
			"action":   actionReference,
			"bindings": bindingReferences,
		},
		map[string]any{
			"kind":         "proposed_use_case",
			"id":           useCaseID,
			"input_ports":  inputs,
			"output_ports": outputs,
			"steps":        []any{stepReference},
		},
	)
	sort.Slice(entities, func(left int, right int) bool {
		leftEntity := entities[left].(map[string]any)
		rightEntity := entities[right].(map[string]any)
		leftKind := leftEntity["kind"].(string)
		rightKind := rightEntity["kind"].(string)
		if leftKind != rightKind {
			return leftKind < rightKind
		}
		return leftEntity["id"].(string) < rightEntity["id"].(string)
	})
	return map[string]any{
		"kind":     "candidate_proposal",
		"root":     useCaseReference,
		"entities": entities,
	}
}

func discoveryPorts(specifications []discoveryPortSpec) []any {
	ports := make([]any, 0, len(specifications))
	for _, specification := range specifications {
		ports = append(ports, map[string]any{
			"kind":       "port",
			"name":       specification.name,
			"value_kind": specification.valueKind,
			"required":   specification.required,
		})
	}
	return ports
}

func discoveryCandidateID(
	candidate map[string]any,
	context map[string]any,
) (string, error) {
	return canonicalLogicalDigest(map[string]any{
		"candidate": map[string]any{
			"kind":       candidate["kind"],
			"subject":    candidate["subject"],
			"evidence":   candidate["evidence"],
			"confidence": candidate["confidence"],
			"proposal":   candidate["proposal"],
		},
		"capability":        context["capability"],
		"inventory_binding": context["inventory_binding"],
		"procedure_uri":     context["procedure_uri"],
		"producer":          context["producer"],
	})
}

func inventoryReferenceForDiscovery(record map[string]any) (map[string]any, error) {
	kind, kindOK := record["kind"].(string)
	identifier, identifierOK := record["id"].(string)
	if !kindOK || !identifierOK {
		return nil, errors.New("inventory reference target is invalid")
	}
	return map[string]any{
		"kind":        "inventory_record_ref",
		"target_kind": kind,
		"target_id":   identifier,
	}, nil
}

func (instance *server) quoteOrderImplementationEvidence(
	records []any,
) ([]any, error) {
	references := make([]any, 0, 10)
	manifestCount := 0
	interfaceCount := 0
	for _, raw := range records {
		record, ok := raw.(map[string]any)
		if !ok {
			return nil, errors.New("quote-order inventory record is invalid")
		}
		switch record["kind"] {
		case "build_manifest":
			if record["dialect_uri"] != goModuleDialectURI {
				return nil, errors.New("quote-order build manifest is unsupported")
			}
			manifestCount++
			references = append(references, inventoryRecordReference(record))
		case "public_interface":
			name, nameOK := record["name"].(string)
			if !nameOK {
				return nil, errors.New("quote-order public interface is invalid")
			}
			if !instance.platformMode {
				_, isOtherFunction := discoveryFunctions[name]
				if isOtherFunction &&
					name != "QuoteOrder" {
					continue
				}
			}
			interfaceCount++
			references = append(references, inventoryRecordReference(record))
		}
	}
	if manifestCount != 1 || interfaceCount != 9 || len(references) != 10 {
		return nil, errors.New("quote-order implementation evidence is incomplete")
	}
	sortDiscoveryReferences(references)
	return references, nil
}

func discoveryEntityReference(kind string, identifier string) map[string]any {
	return map[string]any{
		"kind":        "proposal_entity_ref",
		"target_kind": kind,
		"target_id":   identifier,
	}
}

func discoveryCapability() map[string]any {
	return map[string]any{
		"kind":    "capability",
		"name":    "org.ucf.adapter.discovery",
		"version": discoveryVersion,
	}
}

func sortInventoryRecords(records []map[string]any) {
	sort.Slice(records, func(left int, right int) bool {
		return records[left]["id"].(string) < records[right]["id"].(string)
	})
}

func sortDiscoveryReferences(references []any) {
	sort.Slice(references, func(left int, right int) bool {
		leftReference := references[left].(map[string]any)
		rightReference := references[right].(map[string]any)
		leftKind := leftReference["target_kind"].(string)
		rightKind := rightReference["target_kind"].(string)
		if leftKind != rightKind {
			return leftKind < rightKind
		}
		return leftReference["target_id"].(string) <
			rightReference["target_id"].(string)
	})
}

func sameCanonicalLogicalValue(left any, right any) bool {
	leftEncoded, leftOK := appendCanonicalJSON(nil, left)
	rightEncoded, rightOK := appendCanonicalJSON(nil, right)
	return leftOK && rightOK && bytes.Equal(leftEncoded, rightEncoded)
}

func zeroDigest() string {
	return "0000000000000000000000000000000000000000000000000000000000000000"
}

File: tests/fixtures/adapters/inventory_reference/discovery.py
from __future__ import annotations

import hashlib

from .profile import (
    INVENTORY_SCHEMA_URI,
    INVENTORY_VERSION,
    PRODUCER,
    InvalidProfile,
    canonical_json,
    decode_tagged,
    encode_tagged,
)

DISCOVERY_CAPABILITY = "org.ucf.adapter.discovery"
DISCOVERY_REQUEST_SCHEMA_URI = (
    "urn:ucf:adapter:discovery-request:1.0.0"
)
DISCOVERY_RESULT_SCHEMA_URI = (
    "urn:ucf:adapter:discovery-result:1.0.0"
)
DISCOVERY_PROCEDURE_URI = (
    "urn:ucf:onboarding-procedure:python-ast-discovery:1.0.0"
)
DISCOVERY_CONFIDENCE_BASIS = (
    "urn:ucf:onboarding-confidence:python-public-function:1.0.0"
)
PYTHON_INTERFACE_DIALECT = (
    "urn:ucf:inventory-interface:python-function:1.0.0"
)

_FUNCTIONS = {
    "format_receipt": {
        "slug": "format-receipt",
        "confidence": "0.82",
        "inputs": (("total-cents", "integer", True),),
        "outputs": (("receipt", "string", True),),
    },
    "legacy_discount_hint": {
        "slug": "legacy-discount-hint",
        "confidence": "0.61",
        "inputs": (("code", "string", True),),
        "outputs": (("discount-percent", "integer", False),),
    },
    "normalize_coupon": {
        "slug": "normalize-coupon",
        "confidence": "0.82",
        "inputs": (("code", "string", True),),
        "outputs": (("normalized-code", "string", True),),
    },
    "quote_order": {
        "slug": "quote-order",
        "confidence": "0.82",
        "inputs": (
            ("quantity", "integer", True),
            ("unit-price-cents", "integer", True),
        ),
        "outputs": (("total-cents", "integer", True),),
    },
}


def decode_discovery_request(
    payload: object,
    *,
    inventory: dict[str, object],
) -> dict[str, object]:
    document = _object(payload, "adapter payload")
    _exact(
        document,
        {"kind", "schema_uri", "schema_version", "value"},
        "adapter payload",
    )
    if (
        document["kind"] != "adapter_payload"
        or document["schema_uri"] != DISCOVERY_REQUEST_SCHEMA_URI
        or document["schema_version"] != INVENTORY_VERSION
    ):
        raise InvalidProfile("discovery payload coordinates are incompatible")

    request = _object(
        decode_tagged(document["value"]),
        "discovery request",
    )
    _exact(
        request,
        {
            "kind",
            "onboarding_version",
            "schema_uri",
            "capability",
            "inventory_binding",
            "inventory",
        },
        "discovery request",
    )
    if (
        request["kind"] != "discovery_request_profile"
        or request["onboarding_version"] != INVENTORY_VERSION
        or request["schema_uri"] != DISCOVERY_REQUEST_SCHEMA_URI
    ):
        raise InvalidProfile("discovery request coordinates are incompatible")
    _validate_capability(request["capability"])

    embedded_inventory = _object(
        request["inventory"],
        "embedded inventory",
    )
    if canonical_json(embedded_inventory) != canonical_json(inventory):
        raise InvalidProfile(
            "discovery request does not embed the inventoried snapshot"
        )
    binding = _object(
        request["inventory_binding"],
        "inventory binding",
    )
    _validate_inventory_binding(binding, inventory)
    return request


def build_discovery_result(
    request: dict[str, object],
) -> dict[str, object]:
    inventory = _object(request["inventory"], "embedded inventory")
    records = _array(inventory["records"], "inventory records")
    interfaces = [
        _object(record, "inventory record")
        for record in records
        if (
            type(record) is dict
            and record.get("kind") == "public_interface"
        )
    ]
    interfaces.sort(key=lambda record: str(record["id"]))

    context = {
        "inventory_binding": request["inventory_binding"],
        "producer": PRODUCER,
        "capability": _capability(),
        "procedure_uri": DISCOVERY_PROCEDURE_URI,
    }
    candidates: list[dict[str, object]] = []
    eligible: list[dict[str, object]] = []
    uncovered: list[dict[str, object]] = []
    seen_names: dict[str, int] = {}
    for interface in interfaces:
        name = interface.get("name")
        if type(name) is str:
            seen_names[name] = seen_names.get(name, 0) + 1
    for interface in interfaces:
        reference = _inventory_reference(interface)
        eligible.append(reference)
        name = interface.get("name")
        specification = (
            _FUNCTIONS.get(name)
            if (
                type(name) is str
                and seen_names.get(name) == 1
                and interface.get("interface_kind_uri")
                == PYTHON_INTERFACE_DIALECT
            )
            else None
        )
        if specification is None:
            uncovered.append(reference)
            continue
        proposal = _proposal(specification)
        candidate = {
            "kind": "discovery_candidate",
            "id": "candidate." + ("0" * 64),
            "semantic_digest": _digest(canonical_json(proposal)),
            "subject": reference,
            "evidence": [reference],
            "confidence": {
                "kind": "confidence",
                "scale": "decimal-0-to-1",
                "value": specification["confidence"],
                "basis": DISCOVERY_CONFIDENCE_BASIS,
            },
            "proposal": proposal,
        }
        candidate["id"] = _candidate_id(candidate, context)
        candidates.append(candidate)

    eligible.sort(key=_reference_key)
    uncovered.sort(key=_reference_key)
    candidates.sort(key=lambda candidate: str(candidate["id"]))
    return {
        "kind": "discovery_result_profile",
        "onboarding_version": INVENTORY_VERSION,
        "schema_uri": DISCOVERY_RESULT_SCHEMA_URI,
        **context,
        "coverage": {
            "kind": "discovery_coverage",
            "status": "partial" if uncovered else "complete",
            "eligible_subjects": eligible,
            "uncovered_subjects": uncovered,
        },
        "diagnostics": [],
        "candidates": candidates,
    }


def encode_discovery_result_payload(
    result: dict[str, object],
    *,
    schema_uri: str = DISCOVERY_RESULT_SCHEMA_URI,
) -> dict[str, object]:
    return {
        "kind": "adapter_payload",
        "schema_uri": schema_uri,
        "schema_version": INVENTORY_VERSION,
        "value": encode_tagged(result),
    }


def _validate_capability(value: object) -> None:
    capability = _object(value, "discovery capability")
    _exact(capability, {"kind", "name", "version"}, "capability")
    if capability != _capability():
        raise InvalidProfile("discovery capability is incompatible")


def _validate_inventory_binding(
    binding: dict[str, object],
    inventory: dict[str, object],
) -> None:
    _exact(
        binding,
        {
            "kind",
            "schema_uri",
            "inventory_version",
            "subject_uri",
            "source_revision",
            "canonical_digest",
        },
        "inventory binding",
    )
    expected = {
        "kind": "inventory_binding",
        "schema_uri": INVENTORY_SCHEMA_URI,
        "inventory_version": INVENTORY_VERSION,
        "subject_uri": inventory["subject_uri"],
        "source_revision": inventory["source_revision"],
        "canonical_digest": _digest(canonical_json(inventory)),
    }
    if binding != expected:
        raise InvalidProfile(
            "inventory binding does not name the exact inventoried snapshot"
        )


def _proposal(specification: dict[str, object]) -> dict[str, object]:
    slug = str(specification["slug"])
    action_id = f"action.{slug}"
    step_id = f"step.{slug}"
    use_case_id = f"use-case.{slug}"
    action_reference = _entity_reference("proposed_action", action_id)
    step_reference = _entity_reference("proposed_step", step_id)
    use_case_reference = _entity_reference(
        "proposed_use_case",
        use_case_id,
    )
    inputs = [
        _port(name, value_kind, required)
        for name, value_kind, required in specification["inputs"]
    ]
    outputs = [
        _port(name, value_kind, required)
        for name, value_kind, required in specification["outputs"]
    ]
    bindings: list[dict[str, object]] = []
    binding_references: list[dict[str, object]] = []
    for port in inputs:
        binding_id = f"binding.{slug}.{port['name']}"
        binding_reference = _entity_reference(
            "proposed_binding",
            binding_id,
        )
        binding_references.append(binding_reference)
        bindings.append(
            {
                "kind": "proposed_binding",
                "id": binding_id,
                "target": {
                    "kind": "proposal_port_ref",
                    "owner": step_reference,
                    "direction": "input",
                    "name": port["name"],
                },
                "source": {
                    "kind": "proposal_port_ref",
                    "owner": use_case_reference,
                    "direction": "input",
                    "name": port["name"],
                },
            }
        )
    entities = [
        {
            "kind": "proposed_action",
            "id": action_id,
            "input_ports": inputs,
            "output_ports": outputs,
        },
        *bindings,
        {
            "kind": "proposed_step",
            "id": step_id,
            "action": action_reference,
            "bindings": binding_references,
        },
        {
            "kind": "proposed_use_case",
            "id": use_case_id,
            "input_ports": inputs,
            "output_ports": outputs,
            "steps": [step_reference],
        },
    ]
    entities.sort(key=lambda entity: (str(entity["kind"]), str(entity["id"])))
    return {
        "kind": "candidate_proposal",
        "root": use_case_reference,
        "entities": entities,
    }


def _candidate_id(
    candidate: dict[str, object],
    context: dict[str, object],
) -> str:
    projected_candidate = {
        key: value
        for key, value in candidate.items()
        if key not in {"id", "semantic_digest"}
    }
    projection = {
        "candidate": projected_candidate,
        "capability": context["capability"],
        "inventory_binding": context["inventory_binding"],
        "procedure_uri": context["procedure_uri"],
        "producer": context["producer"],
    }
    return "candidate." + hashlib.sha256(
        canonical_json(projection)
    ).hexdigest()


def _inventory_reference(
    interface: dict[str, object],
) -> dict[str, object]:
    identifier = interface.get("id")
    if type(identifier) is not str:
        raise InvalidProfile("public interface identifier is invalid")
    return {
        "kind": "inventory_record_ref",
        "target_kind": "public_interface",
        "target_id": identifier,
    }


def _entity_reference(
    kind: str,
    identifier: str,
) -> dict[str, object]:
    return {
        "kind": "proposal_entity_ref",
        "target_kind": kind,
        "target_id": identifier,
    }


def _port(
    name: str,
    value_kind: str,
    required: bool,
) -> dict[str, object]:
    return {
        "kind": "port",
        "name": name,
        "value_kind": value_kind,
        "required": required,
    }


def _capability() -> dict[str, str]:
    return {
        "kind": "capability",
        "name": DISCOVERY_CAPABILITY,
        "version": INVENTORY_VERSION,
    }


def _digest(payload: bytes) -> dict[str, str]:
    return {
        "kind": "digest",
        "algorithm": "sha-256",
        "value": hashlib.sha256(payload).hexdigest(),
    }


def _reference_key(reference: dict[str, object]) -> tuple[str, str]:
    return (
        str(reference["target_kind"]),
        str(reference["target_id"]),
    )


def _exact(
    value: dict[str, object],
    expected: set[str],
    label: str,
) -> None:
    if set(value) != expected:
        raise InvalidProfile(f"{label} fields are not exact")


def _object(value: object, label: str) -> dict[str, object]:
    if type(value) is not dict:
        raise InvalidProfile(f"{label} must be an object")
    return value


def _array(value: object, label: str) -> list[object]:
    if type(value) is not list:
        raise InvalidProfile(f"{label} must be a list")
    return value
